### PR TITLE
Azure Provider: Authenticating using the Azure CLI

### PR DESCRIPTION
**Pull request by v-mavick**
First editorial pass. Feedback greatly welcome!
@tombuildsstuff: Should the references to `terraform` and `az` on line 22 instead be Terraform and Azure?

### CODE REVIEW

**Comment by tombuildsstuff** [comment]
hey @v-mavick 

Thanks for this PR :)

> Should the references to terraform and az on line 22 instead be Terraform and Azure?

Nope - taking a look at this these are a reference to the binaries `terraform` / `az` - so I think these are fine as-is (that said, we could probably rephrase this sentence without the reference? 🤔) 

Thanks!

**Comment by v-mavick** [comment]
Thanks! That’s what I suspected, but I wasn’t sure.

Marc Vick
Sent from my iPhone

On Nov 13, 2018, at 12:48 AM, Tom Harvey <notifications@github.com<mailto:notifications@github.com>> wrote:


hey @v-mavick<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fv-mavick&data=02%7C01%7Cv-mavick%40microsoft.com%7C0f520d46fa714c7073ac08d64944d032%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636776957236973697&sdata=WIIewtqOM2W%2BAd2vjBDsksJu9%2B%2BIwGd08t7X8wuRh0I%3D&reserved=0>

Thanks for this PR :)

Should the references to terraform and az on line 22 instead be Terraform and Azure?

Nope - taking a look at this these are a reference to the binaries terraform / az - so I think these are fine as-is (that said, we could probably rephrase this sentence without the reference? 🤔)

Thanks!

File: website/docs/authenticating_via_azure_cli.html.markdown [hunk 1]
```diff
@@ -9,77 +9,81 @@ description: |-
...
-```
+    ```shell
+    $ az login
+    ```
 
-~> **NOTE:** Authenticating via the Azure CLI is only supported when using a User Account. If you're using a Service Principal (e.g. via `az login --service-principal`) you should instead [authenticate via the Service Principal directly](authenticating_via_service_principal.html).
+    ~> **NOTE:** Authenticating via the Azure CLI is only supported when using a User Account. If you're using a Service Principal (for example via `az login --service-principal`), you should instead [authenticate via the Service Principal directly](authenticating_via_service_principal.html).
 
-This will prompt you to open a web browser, as shown below:
+1. You are prompted to open a web browser:
```
Comment (tombuildsstuff) [inline website/docs/authenticating_via_azure_cli.html.markdown]: does this want to be `2.`?

File: website/docs/authenticating_via_azure_cli.html.markdown [hunk 2]
```diff
@@ -9,77 +9,81 @@ description: |-
...
 
-```shell
-To sign in, use a web browser to open the page https://aka.ms/devicelogin and enter the code XXXXXXXX to authenticate.
-```
+    ```shell
+    To sign in, use a web browser to open the page https://aka.ms/devicelogin and enter the code XXXXXXXX to authenticate.
+    ```
 
-Once logged in - it's possible to list the Subscriptions associated with the account via:
+1. Once logged in, it's possible to list the Subscriptions associated with the account via:
```
Comment (tombuildsstuff) [inline website/docs/authenticating_via_azure_cli.html.markdown]: does this want to be `3.`?

**Comment by v-mavick** [comment]
Great. I’ll get back on this first thing in the morning.

Marc Vick
Sent from my iPhone

On Nov 13, 2018, at 1:02 AM, Tom Harvey <notifications@github.com<mailto:notifications@github.com>> wrote:


@tombuildsstuff requested changes on this pull request.

hey @v-mavick<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fv-mavick&data=02%7C01%7Cv-mavick%40microsoft.com%7C657d48881d5b444df8a808d64946bc2f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636776965490558060&sdata=px8ZZ4j6XQgTHhVzpZ1Dy9Yz1zIJgYfzZ7Ln3jiXt4c%3D&reserved=0>

Thanks for this PR :)

On the whole this is looking pretty good - if we can fix up the minor issues (and sort out the indentation) this should otherwise be good to merge 👍

Thanks!

________________________________

In website/docs/authenticating_via_azure_cli.html.markdown<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fterraform-providers%2Fterraform-provider-azurerm%2Fpull%2F2300%23discussion_r232943206&data=02%7C01%7Cv-mavick%40microsoft.com%7C657d48881d5b444df8a808d64946bc2f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636776965490568065&sdata=4frOWVQ9WysUxxFOmZMjbjU1yjYP9E%2BaQgNXXnvXEck%3D&reserved=0>:

>
-  When `terraform`  and `az` are run on hosts / containers with different timezones, the variable $TZ should be set on the host.
+When `terraform` and `az` are run on hosts/containers with different timezones, the variable $TZ should be set on the host.


I feel like line 20 / 22 are saying the same thing; so perhaps we could combine them? 🤔

What do you think of:

When the timezones in which az and terraform differ, such as when these binaries are run from within Docker (which defaults to UTC) and the system timezone (which may not) - Terraform may incorrectly detect these access tokens as expired. It's possible to work around this by ensuring the Environment Variable $TZ is configured the same on both.

________________________________

In website/docs/authenticating_via_azure_cli.html.markdown<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fterraform-providers%2Fterraform-provider-azurerm%2Fpull%2F2300%23discussion_r232943605&data=02%7C01%7Cv-mavick%40microsoft.com%7C657d48881d5b444df8a808d64946bc2f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636776965490568065&sdata=lKZ35yOHJcByc5X12Jw0IhAyQj9Kzsnpb3LPJfD8D3c%3D&reserved=0>:

>
 ## Configuring the Azure CLI

-~> **Note:** There are multiple versions of the Azure CLI - the latest version is known as [the Azure CLI 2.0 (Python)](https://github.com/Azure/azure-cli) and [the older Azure CLI (Node.JS)](https://github.com/Azure/azure-xplat-cli). While Terraform currently supports both - we highly recommend users upgrade to the Azure CLI 2.0 (Python) if possible.
+~> **Note:** There are multiple versions of the Azure CLI. The latest version is known as [the Azure CLI 2.0 (Python)](https://github.com/Azure/azure-cli) and [the older Azure CLI (Node.JS)](https://github.com/Azure/azure-xplat-cli). While Terraform currently supports both, we highly recommend users upgrade to the Azure CLI 2.0 (Python), if possible.


in retrospect I think it's worth making highly recommend bold?

⬇️ Suggested change

-~> **Note:** There are multiple versions of the Azure CLI. The latest version is known as [the Azure CLI 2.0 (Python)](https://github.com/Azure/azure-cli) and [the older Azure CLI (Node.JS)](https://github.com/Azure/azure-xplat-cli). While Terraform currently supports both, we highly recommend users upgrade to the Azure CLI 2.0 (Python), if possible.
+~> **Note:** There are multiple versions of the Azure CLI. The latest version is known as [the Azure CLI 2.0 (Python)](https://github.com/Azure/azure-cli) and [the older Azure CLI (Node.JS)](https://github.com/Azure/azure-xplat-cli). While Terraform currently supports both, we **highly recommend** users upgrade to the Azure CLI 2.0 (Python), if possible.


________________________________

In website/docs/authenticating_via_azure_cli.html.markdown<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fterraform-providers%2Fterraform-provider-azurerm%2Fpull%2F2300%23discussion_r232943787&data=02%7C01%7Cv-mavick%40microsoft.com%7C657d48881d5b444df8a808d64946bc2f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636776965490578074&sdata=fN1qTyqhzriTgriYNCbnxcQjnolDcbE9rfqc7DnRBu0%3D&reserved=0>:

>
-```shell
-$ az login
-```
+    ```shell
+    $ az login
+    ```


can we remove this indentation?

________________________________

In website/docs/authenticating_via_azure_cli.html.markdown<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fterraform-providers%2Fterraform-provider-azurerm%2Fpull%2F2300%23discussion_r232944098&data=02%7C01%7Cv-mavick%40microsoft.com%7C657d48881d5b444df8a808d64946bc2f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636776965490578074&sdata=2az%2FBxFIK4Ic%2BmYWFuECj281WYNQ5BQnQShwywUuGkg%3D&reserved=0>:

>
-```shell
-To sign in, use a web browser to open the page https://aka.ms/devicelogin and enter the code XXXXXXXX to authenticate.
-```
+    ```shell
+    To sign in, use a web browser to open the page https://aka.ms/devicelogin and enter the code XXXXXXXX to authenticate.
+    ```


this behaviours changed slightly since then, so it could be worth updating the output from this block?

________________________________

In website/docs/authenticating_via_azure_cli.html.markdown<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fterraform-providers%2Fterraform-provider-azurerm%2Fpull%2F2300%23discussion_r232944383&data=02%7C01%7Cv-mavick%40microsoft.com%7C657d48881d5b444df8a808d64946bc2f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636776965490588079&sdata=Sewkhe3oiqnmoTv5WYgESGEtJfY%2BeSKeMVo5unzJA%2BE%3D&reserved=0>:

>
-Also, if you have been authenticating with a service principal and you switch to Azure CLI, you must null out the ARM_* environment variables. Failure to do so causes errors to be thrown.
+    Also, if you have been authenticating with a service principal and you switch to Azure CLI, you must null out the ARM_* environment variables. Failure to do so causes errors to be thrown.


can we remove this indentation? this'll mean it's not rendered correctly, unfortunately

________________________________

In website/docs/authenticating_via_azure_cli.html.markdown<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fterraform-providers%2Fterraform-provider-azurerm%2Fpull%2F2300%23discussion_r232946401&data=02%7C01%7Cv-mavick%40microsoft.com%7C657d48881d5b444df8a808d64946bc2f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636776965490588079&sdata=XrDFZ7tL6pKNCCyCdmMlGOz2Gp2mvUWFNxcIHb1V0vg%3D&reserved=0>:

>
-This will prompt you to open a web browser, as shown below:
+1. You are prompted to open a web browser:


does this want to be 2.?

________________________________

In website/docs/authenticating_via_azure_cli.html.markdown<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fterraform-providers%2Fterraform-provider-azurerm%2Fpull%2F2300%23discussion_r232946454&data=02%7C01%7Cv-mavick%40microsoft.com%7C657d48881d5b444df8a808d64946bc2f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636776965490598084&sdata=WsLJiLEJjYxGfWP%2B4Hm2TMu%2BljFwjHigJvOxWuHxxCw%3D&reserved=0>:

>
-Once logged in - it's possible to list the Subscriptions associated with the account via:
+1. Once logged in, it's possible to list the Subscriptions associated with the account via:


does this want to be 3.?

File: website/docs/authenticating_via_azure_cli.html.markdown [hunk 3]
```diff
@@ -9,77 +9,81 @@ description: |-
...
+~> **NOTE:** When `az login` fetches the access tokens, they are interpreted (and stored) according to the timezone settings in which the Azure CLI is running.
   
-  When the timezones `az` and `terraform` run in differ (for example when `az` is run inside docker, which defaults to UTC, and the system timezone where `terraform` runs is not UTC), `terraform` interprets the token differently from what `az` intended and may incorrectly determine the token to be stale and invalid.
+When the timezones in which `az` and `terraform` run differ from each other, `terraform` interprets the token differently from what `az` intended, and may incorrectly determine the token to be stale and invalid. An example of this is when `az` is run inside Docker (which defaults to UTC) and the system timezone in which `terraform` runs is not UTC.
   
-  When `terraform`  and `az` are run on hosts / containers with different timezones, the variable $TZ should be set on the host.  
+When `terraform` and `az` are run on hosts/containers with different timezones, the variable $TZ should be set on the host.
```
[the remark this answers is not in the document]
Comment (v-mavick) [reply website/docs/authenticating_via_azure_cli.html.markdown]: Modified as suggested.

File: website/docs/authenticating_via_azure_cli.html.markdown [hunk 4]
```diff
@@ -9,77 +9,81 @@ description: |-
...
 
-Firstly, login to the Azure CLI using:
+1. Log in to the Azure CLI using:
 
-```shell
-$ az login
-```
+    ```shell
+    $ az login
+    ```
```
[the remark this answers is not in the document]
Comment (v-mavick) [reply website/docs/authenticating_via_azure_cli.html.markdown]: Removed indentation.

[the remark this answers is not in the document]
Comment (v-mavick) [reply website/docs/authenticating_via_azure_cli.html.markdown]: Note: Removing the indents on these code examples destroys the automatic numbering (1-4) of these steps.

File: website/docs/authenticating_via_azure_cli.html.markdown [hunk 5]
```diff
@@ -9,77 +9,81 @@ description: |-
...
 
-This will prompt you to open a web browser, as shown below:
+1. You are prompted to open a web browser:
 
-```shell
-To sign in, use a web browser to open the page https://aka.ms/devicelogin and enter the code XXXXXXXX to authenticate.
-```
+    ```shell
+    To sign in, use a web browser to open the page https://aka.ms/devicelogin and enter the code XXXXXXXX to authenticate.
+    ```
```
[the remark this answers is not in the document]
Comment (v-mavick) [reply website/docs/authenticating_via_azure_cli.html.markdown]: Is this something I need to address in this document?

File: website/docs/authenticating_via_azure_cli.html.markdown [hunk 6]
```diff
@@ -9,77 +9,81 @@ description: |-
...
+
+    ~> **Note:** When authenticating via the Azure CLI, Terraform will automatically connect to the Default Subscription. Therefore, if you have multiple subscriptions on the account, you may need to set the correct Default Subscription, via:
+
+    ```shell
+    $ az account set --subscription="SUBSCRIPTION_ID"
+    ```
 
-Also, if you have been authenticating with a service principal and you switch to Azure CLI, you must null out the ARM_* environment variables. Failure to do so causes errors to be thrown.
+    Also, if you have been authenticating with a service principal and you switch to Azure CLI, you must null out the ARM_* environment variables. Failure to do so causes errors to be thrown.
```
[the remark this answers is not in the document]
Comment (v-mavick) [reply website/docs/authenticating_via_azure_cli.html.markdown]: Indentation removed.

File: website/docs/authenticating_via_azure_cli.html.markdown (hunk 1 above)
Comment (v-mavick) [reply website/docs/authenticating_via_azure_cli.html.markdown]: Yes. It should be corrected in the next commit.

File: website/docs/authenticating_via_azure_cli.html.markdown (hunk 2 above)
Comment (v-mavick) [reply website/docs/authenticating_via_azure_cli.html.markdown]: Yes. I "hard-coded" the numbers, which should all display correctly now.

**Comment by tombuildsstuff** [comment]
hey @v-mavick 

Thanks for pushing those changes - there's some changes coming to Azure CLI authentication in the form of #2387 such that this document changes somewhat. Since #2387 changes the document slightly (insofar as we no longer parse the Azure CLI access tokens) I've incorporated the changes made in this document into that PR - I hope you don't mind!

Thanks!

**Comment by v-mavick** [comment]
Tom,

Thanks for the update. That may explain why I was having issues pushing to #2300!

**Pull request closed without merging**